homeassistant/components/switch/zha.py

```diff
@@ -65,23 +68,29 @@ def is_on(self) -> bool:
         """Turn the entity on."""
         from zigpy.exceptions import DeliveryError
         try:
-            await self._endpoint.on_off.on()
+            res = await self._endpoint.on_off.on()
+            _LOGGER.debug("%s: turned 'on': %s", self.entity_id, res[1])
         except DeliveryError as ex:
-            _LOGGER.error("Unable to turn the switch on: %s", ex)
+            _LOGGER.error("%s: Unable to turn the switch on: %s",
+                          self.entity_id, ex)
             return
 
         self._state = 1
+        self.async_schedule_update_ha_state()
```

Comment:
We shouldn't do this, and neither should we change `self._state`. We depend on `attribute_updated` to push the state change to us

Reply:
I'm not entirely sure relying on `attribute_updated` is a proper channel for this kind of state updated, especially when we're triggering the change. 
We do send a ZCL command and we do know the transmission hasn't timed out, in other words device has acknowledge reception of the command. 
The problem with `attribute_updated` it relies on attribute reporting configuration, so if for whatever reason `zha.configure_reporting()` was unsuccessful, this makes the switch unusable, as I never can change the state (because the feedback loop through `attribute_updated` is broken) even though I can control the switch and switch can successfully receive my commands. 
`attribute_updated` channel is more for externally  originated events, like user toggling the switch from the button on the actual physical device. 

@rcloran what do you think?

Reply:
I also want to implement `attribute_updated()` method for `fan.zha.ZhaFan` class, because it is nice for HA state to be in sync with the actual state of the device, since the fan is often operated externally from the remote/wall switch.
I'd like to understand why it is fundamentally wrong with updating `self._state` in the service calls and in `attribute_updated()` method. Both are executed in the event loop, so there shouldn't be any race conditions. I'm not always getting the `attribute_update()` with King of Fans zigbee fan, so relying solely on `atribute_updated()` to update `self._state` would likely render Fan inoperable from HA